src/plugins/intel_cpu/src/nodes/tensoriterator.cpp

```diff
@@ -231,97 +231,148 @@ DynamicBuffer::DynamicBuffer(const MemoryPtr &from_, const std::vector<MemoryPtr
-    const auto abs_stride = std::abs(stride);
-
-    const auto old_desc = mem_holder_buffer.get_desc();
-    auto dims = old_desc.dims();
-
-    if (from->getStaticDims()[axis] != abs_stride)
-        IE_THROW() << "TensorIterator (Loop) has incorrect output shape[axis] after iteration for concatenation. " << abs_stride <<
-        " is expected, but actual: " << from->getStaticDims()[axis];
+    if (!mem_holder_buffer) { // else reuse buffer holder of last inference
+        // preallocate a large chunk of memory to hold intermediate concated outputs of all iterations.
+        mem_holder_buffer = create_buffer(eng);
+    }
 
-    dims[axis] += abs_stride;
-    dnnl::memory::desc new_buffer_desc(dims, old_desc.data_type(), DnnlExtensionUtils::GetPlainFormatByRank(dims.size()));
+    // reset chunk_offset_in_byte since the first execution
+    chunk_stride_in_byte = mem_holder_buffer->GetSize() / count;
+    chunk_offset_in_byte = stride > 0 ? 0 : (chunk_stride_in_byte - chunk_unit_in_byte);
+    num_execs = 0;
+}
 
-    if (stride > 0.0f) {
-        chunk_offset_in_byte += new_buffer_desc.data.format_desc.blocking.strides[axis] * elem_size * abs_stride;
+bool DynamicBuffer::check_buffer() {
+    if (map_rule.stride > 0) {
+        if (chunk_offset_in_byte + chunk_unit_in_byte > chunk_stride_in_byte) return true;
     } else {
-        buffer_offset_in_byte = from->GetPrimitive().get_desc().data.format_desc.blocking.strides[axis] * elem_size * abs_stride;
+        if (chunk_offset_in_byte < 0) return true;
     }
+    return false;
+}
+
+MemoryPtr DynamicBuffer::create_buffer(const dnnl::engine& eng) {
+    const auto abs_stride = std::abs(map_rule.stride);
 
-    return dnnl::memory(new_buffer_desc, eng);
+    const auto estimate_iters = [&] () {
+        if (max_iter_count != -1) return max_iter_count;
+
+        // in case of no idea of memory upper boundary
+        return (num_execs == 0) ? 1 : 2 * num_execs; // growth factor 2
+    };
+    const auto estimated_iters = estimate_iters();
+    const Shape _shape = Shape({count, static_cast<size_t>(abs_stride * estimated_iters), len/elem_size});
+    auto _descCreator = BlockedDescCreator::getCommonCreators().at(LayoutType::ncsp);
+    auto new_buffer_desc = _descCreator->createSharedDesc(from->getDesc().getPrecision(), _shape);
+
+    auto _ptr = MemoryPtr(new Memory(eng));
+    _ptr->Create(*new_buffer_desc);
+    return _ptr;
 }
 
-void DynamicBuffer::move_buffer(dnnl::memory new_buffer) {
-    const auto axis = map_rule.axis;
-    const auto src_stride = mem_holder_buffer.get_desc().dims()[axis] * len;
-    const auto dst_stride = new_buffer.get_desc().dims()[axis] * len;
+void DynamicBuffer::move_buffer(MemoryPtr& new_buffer) {
+    const auto stride = map_rule.stride;
+
+    // copy data from old buffer to new buffer
+    const auto src_stride = chunk_stride_in_byte;
+    const auto dst_stride = new_buffer->getDesc().as<BlockedMemoryDesc>()->getBlockDims()[1] * len;
```

Comment:
```suggestion
    const auto dst_stride = new_buffer->getStaticDims()[1] * len;
```
Since we use plain format.